pkg/scalers/predictkube_scaler.go

```diff
@@ -0,0 +1,463 @@
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := pks.grpcClient.GetPredictMetric(ctx, &pb.ReqGetPredictMetric{
+		ForecastHorizon: uint64(math.Round(float64(pks.metadata.predictHorizon / pks.metadata.stepDuration))),
+		Observations:    results,
+	})
+
+	if err != nil {
+		return 0, err
+	}
+
+	var y int64
+	if len(results) > 0 {
+		y = int64(results[len(results)-1].Value)
+	}
+
+	x := resp.GetResultMetric()
+
+	return func(x, y int64) int64 {
+		if x < y {
+			return y
+		}
+		return x
+	}(x, y), nil
+}
+
+func (pks *PredictKubeScaler) doQuery(ctx context.Context) ([]*commonproto.Item, error) {
+	currentTime := time.Now().UTC()
+
+	if pks.metadata.stepDuration == 0 {
+		pks.metadata.stepDuration = defaultStep
+	}
+
+	r := v1.Range{
+		Start: currentTime.Add(-pks.metadata.historyTimeWindow),
+		End:   currentTime,
+		Step:  pks.metadata.stepDuration,
+	}
+
+	val, warns, err := pks.api.QueryRange(ctx, pks.metadata.query, r)
+
+	if len(warns) > 0 {
+		predictKubeLog.V(1).Info("warnings", warns)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return pks.parsePrometheusResult(val)
+}
+
+func (pks *PredictKubeScaler) parsePrometheusResult(result model.Value) (out []*commonproto.Item, err error) {
+	metricName := GenerateMetricNameWithIndex(pks.metadata.scalerIndex, kedautil.NormalizeString(fmt.Sprintf("predictkube-%s", pks.metadata.metricName)))
+	switch result.Type() {
+	case model.ValVector:
+		if res, ok := result.(model.Vector); ok {
+			for _, val := range res {
+				t, err := tc.AdaptTimeToPbTimestamp(tc.TimeToTimePtr(val.Timestamp.Time()))
+				if err != nil {
+					return nil, err
+				}
+
+				out = append(out, &commonproto.Item{
+					Timestamp:  t,
+					Value:      float64(val.Value),
+					MetricName: metricName,
+				})
+			}
+		}
+	case model.ValMatrix:
+		if res, ok := result.(model.Matrix); ok {
+			for _, val := range res {
+				for _, v := range val.Values {
+					t, err := tc.AdaptTimeToPbTimestamp(tc.TimeToTimePtr(v.Timestamp.Time()))
+					if err != nil {
+						return nil, err
+					}
+
+					out = append(out, &commonproto.Item{
+						Timestamp:  t,
+						Value:      float64(v.Value),
+						MetricName: metricName,
+					})
+				}
+			}
+		}
+	case model.ValScalar:
+		if res, ok := result.(*model.Scalar); ok {
+			t, err := tc.AdaptTimeToPbTimestamp(tc.TimeToTimePtr(res.Timestamp.Time()))
+			if err != nil {
+				return nil, err
+			}
+
+			out = append(out, &commonproto.Item{
+				Timestamp:  t,
+				Value:      float64(res.Value),
+				MetricName: metricName,
+			})
+		}
+	case model.ValString:
+		if res, ok := result.(*model.String); ok {
+			t, err := tc.AdaptTimeToPbTimestamp(tc.TimeToTimePtr(res.Timestamp.Time()))
+			if err != nil {
+				return nil, err
+			}
+
+			s, err := strconv.ParseFloat(res.Value, 64)
+			if err != nil {
+				return nil, err
+			}
+
+			out = append(out, &commonproto.Item{
+				Timestamp:  t,
+				Value:      s,
+				MetricName: metricName,
+			})
+		}
+	}
+
+	return out, nil
+}
```

Comment:
I don't get the objective for this method. 
Could you explain it to me? 
I don't understand either why you need the metricName here, I mean, the metricName depends on the index inside the ScaledObject and it is deterministic but it also could change if the user add/remove a trigger

Reply:
@JorTurFer In few words, with this method we parse answer from Prometheus server. @zroubalik done.